### PR TITLE
updated stageGroup validation scripts so clinical/pathological_tumour…

### PR DESCRIPTION
**Pull request by hknahal**
…_staging_system is required if clinical/pathological_stage_group are submitted. Also updated primary_diagnosis schema with re-ordered clinical staging fields so they are consistent.
Related to https://github.com/icgc-argo/argo-dictionary/issues/229 and https://github.com/icgc-argo/argo-dictionary/issues/228

### CODE REVIEW

**Review by rosibaj — CHANGES_REQUESTED** [review]
1) I documented a separate bug which i noticed, related to this area:   https://github.com/icgc-argo/argo-dictionary/issues/234

This is confirmed to be happening with these changes as well, so i think its something we want to fix right now as well. 

2) A question about your implementation.  The other validation is cpatured in the /common validations folder as it is used for both files (pd + specimen).  Do you see us in the future adding the same set of staging groupings to even more files? if so, then should the script also be generalized as a /common script used across files?  If not, then i think that this is working fine and will get an approve from me!

File: references/validationFunctions/primary_diagnosis/stageGroup.js [hunk 1]
```diff
@@ -28,7 +28,12 @@ const validation = () =>
   (function validate(inputs) {
     const {$row, $name, $field} = inputs;
     let result = { valid: true, message: 'Ok' };
-    if ($row.clinical_tumour_staging_system && $field) {
+    if (!($row.clinical_tumour_staging_system) && $field) {
+       result.valid = false;
+       const msg = "'clinical_tumour_staging_system' must be submitted if 'clinical_stage_group' is submitted";
```
Comment (rosibaj) [inline references/validationFunctions/primary_diagnosis/stageGroup.js:33]: Friendlier error for ui: 
`The field` 'clinical_tumour_staging_system' must be submitted if 'clinical_stage_group' is submitted"`.`

File: references/validationFunctions/specimen/stageGroup.js [hunk 2]
```diff
@@ -28,7 +28,12 @@ const validation = () =>
   (function validate(inputs) {
     const {$row, $name, $field} = inputs;
     let result = { valid: true, message: 'Ok' };
-    if ($row.pathological_tumour_staging_system && $field) {
+    if (!($row.pathological_tumour_staging_system) && $field) {
+       result.valid = false;
+       const msg = "'pathological_tumour_staging_system' must be submitted if 'pathological_stage_group' is submitted";
```
Comment (rosibaj) [inline references/validationFunctions/specimen/stageGroup.js:33]: `The field` 'pathological_tumour_staging_system' must be submitted if 'pathological_stage_group' is submitted"`.`

**Comment by hknahal** [comment]
> 1. I documented a separate bug which i noticed, related to this area:   #234

@rosibaj I fixed this issue in https://github.com/icgc-argo/argo-dictionary/issues/234

> 2. A question about your implementation.  The other validation is cpatured in the /common validations folder as it is used for both files (pd + specimen).  Do you see us in the future adding the same set of staging groupings to even more files? if so, then should the script also be generalized as a /common script used across files?  If not, then i think that this is working fine and will get an approve from me!

@rosibaj - Good suggestion. I merged the primary diagnosis, specimen and followup stageGroup validation code into one script under the /common validations folder.

**Review by rosibaj — APPROVED** [review]

**Comment by rosibaj** [comment]
tested on qa
- had to submit the pathological staging field if stage submitted
- could not submit non-permissible values
- could submit just staging field without stage.

**Comment by hknahal** [comment]
> * could submit just staging field without stage.

@rosibaj I think this shouldn't have happened. The stage group field is required if the specimen is tumour. So if the staging system is submitted but the stage group is missing, it should have reported an error - I thought this is how the validation worked before, but I'll investigate.

**Comment by hknahal** [comment]
> * could submit just staging field without stage.

@rosibaj I fixed this issue now. I compiled the dictionary fix and deployed it in QA (v. 80.4). Here's an example entry in the specimen.tsv where the `pathological_staging_system` was submitted, but the `pathological_stage_group` field was empty. This should now invalidate (https://platform-ui.qa.argo.cancercollaboratory.org/submission/program/HNAHAL-CA/clinical-submission?tab=specimen):

|program_id | submitter_donor_id | submitter_specimen_id | submitter_primary_diagnosis_id | pathological_tumour_staging_system | pathological_t_category | pathological_n_category | pathological_m_category | pathological_stage_group | specimen_acquisition_interval | tumour_histological_type | specimen_anatomic_location | specimen_processing | specimen_storage | reference_pathology_confirmed | tumour_grading_system | tumour_grade | percent_tumour_cells | percent_proliferating_cells | percent_inflammatory_tissue | percent_stromal_cells | percent_necrosis|
|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|
|HNAHAL-CA | DASH-6 | dashsp-6.2 | dashpd-6 | Lugano staging system |   |   |   |   | 2 | 9691/36 | C25.1 |   |   | Yes | ISUP grading system | G2 | 0.1 | 0.6 | 0.2 | 0.1 | 0|

**Comment by rosibaj** [comment]
Tested dictionary 80.4 against these test cases: 

```
If sample_registration.tumour_normal_desigantion = Tumour, 
	then specimen.pathological_staging_system OR primary_diagnosis.clinical_staging_system MUST have a value. 
If <pathological or clinical>_staging_system has a value, 
	then <pathological or clinical>_stage_group must have a value AND that value must be selected depending on the value in the staging_system field. 
		AND IF <pathological or clinical>_staging_system = AJCC 
			then <pathological or clinical>_t_value, <pathological or clinical>_n_value, <pathological or clinical>_m_value must be provided. 
If <pathological or clinical>_staging_system does NOT has a value
	then <pathological or clinical>_stage_group, <pathological or clinical>_t_value, <pathological or clinical>_n_value, <pathological or clinical>_m_value must NOT have a value.
```

Results:

- if `stage group` submitted with no `pathological_staging_system` got error saying i need to submit staging system
- if `T/N/M` fields submitted with no `pathological_staging_system` got error saying i need to submit an AJCC option for the staging system
- When `pathological_staging_system= AJCC` was not selected(ex used Binet) , T/N/M were not allowed to be submitted
- When `pathological_staging_system= AJCC` was selected, i was prompted to put in T/N/M correctly depending on which i left blank
- When fields were submitted for a normal sample, validation worked: 
image

@hknahal LGTM!!

**Review by rosibaj — APPROVED** [review]

**Pull request merged**